### ISSUE TITLE
Cloud storage: deal with Dropbox short-lived token.

### ISSUE THREAD

**Issue by anphunl** [koreader/koreader#8571]
* KOReader version: All
* Device: All

#### Issue
On September 30th, 2021, Dropbox retired the creation of long-lived access tokens — resulting in all new tokens being short-lived.
Ref: https://dropbox.tech/developers/migrating-app-permissions-and-access-tokens

#### Steps to reproduce
* Go to the Dropbox Console and generate a new token, it will starts with `sl` (short-lived)
* Use that token within Cloud Storage and it works for now.
* This token will be expired very soon (about a day) and it will show the message "Cannot fetch..."

**Comment by AlexanderTN** [comment]
I have the same problem, the token is expired within 24 hours. Could you please apply Refresh Token for Koreader?
Thank you!

**Comment by Frenzie** [comment]
Quoting @mkozlows relevant reply so it's not lost and forgotten.

@mkozlows See #9193.


https://github.com/koreader/koreader/pull/2424#issuecomment-1140297078 

> So yeah, Dropbox really modernized their auth flows, and deprecated all the legacy ones. Which is, I guess, great for security, but awkward for KOReader.
> 
> Relevant documentation about the Dropbox auth changes:
> 
> [Conceptual guide to supported OAuth flows](https://developers.dropbox.com/oauth-guide) [OAuth API docs](https://www.dropbox.com/developers/documentation/http/documentation#authorization)
> 
> Basically the tldr is that Dropbox needs an OAuth PKCE flow, which means that at the very least, the user would need to go through a web-based signup flow. This would require KOReader to have 1) web client UI to the Dropbox OAuth site, and 2) some way to handle a redirect URL client-side (there's [documentation for this](https://www.oauth.com/oauth2-servers/redirect-uris/redirect-uris-native-apps/) for mobile OSes, but I don't know how applicable it is to KOReader.)
> 
> The approach that would simplify this enormously would be if Dropbox supported [OAuth 2.0 Device Authorization Grant](https://oauth.net/2/device-flow/), which would let the user enter a code into a separate web browser to authorize the device. Unfortunately, [Dropbox doesn't support that](https://www.dropboxforum.com/t5/Dropbox-API-Support-Feedback/Does-Dropbox-support-OAuth-2-0-Device-Authorization-Grant/td-p/397163).
> 
> If you've used the built-in Dropbox integration on certain Kobo devices, and are saying "but those have device flow enabled," the reason is that [Kobo created a Kobo-controlled proxy](https://help.kobo.com/hc/en-us/articles/360033830114-Add-books-to-your-Kobo-Sage-Kobo-Forma-or-Kobo-Elipsa-with-Dropbox) to enable the device flow. That's not reusable by KOReader; it would be possible for someone to create a similar proxy for KOReader to use ([sample repo here](https://github.com/aaronpk/Device-Flow-Proxy-Server)), but that's obviously an awkward fit for an open-source project of this type -- whoever runs that proxy server would effectively have access to whatever permission scope the app used. (Kobo scopes the permission to just a specific app folder, which is fine for Kobo, but...)
> 
> Long story short, this isn't a minor tweak, it's a complete rethinking and major rework of Dropbox integration, unfortunately. Until that's done, it's probably best to think of Dropbox integration as dead.

**Comment by shamanNS** [comment]
Would someone be willing to make necessary changes to this plugin to make it usable at least for users that could manually do the steps needed to get a valid refresh_token. So that the plugin then just needs to be able to send "grant_type=refresh_token" POST request 

`https://api.dropboxapi.com/oauth2/token?grant_type=refresh_token&refresh_token=`


to fetch a short-lived access_token, instead of current behavior of reading the string token from plugin config file? Refresh_token is long lived (I'm not sure just how long lived exactly, this was the only quote I could find: "long-lived refresh_token that can be used to request a new short-lived access token as long as a user's approval remains valid.").

So the plugin UI would expect the user to fill out 3 things:

- app_id
- app_secret
- a valid refresh_token

And then the Lua code would need to send just a single POST request to get valid access_token and use it for all Dropbox API calls.
Admittedly, the new "must be done by user using web browser" part is more involved and more difficult for average users then the old/currently broken approach but is doable and should be a one time thing just like it was previously when Dropbox API used long-lived access_tokens. 



In summary: users job would be to figure out how to generate refresh_token and KOReader's Dropbox plugin would need couple of changes how it acquires a token and would need to keep track of expire times / periodically exchange long-lived refresh_token for short-lived access_token.

**Comment by hius07** [comment]
> So the plugin UI would expect the user to fill out 3 things:

Something like https://github.com/koreader/koreader/pull/9131#issuecomment-1140397070, OK?

**Comment by shamanNS** [comment]
Yes, just like that. 
Sorry, it never crossed my mind to check if there is already a pull request that addresses that issue :)

So, if I understood things correctly that PR will be merged when KOReader 2022.08 is released? (next month?)

**Comment by hius07** [comment]
You can take 3 files from the above PR to try short-lived token access.

**Comment by shamanNS** [comment]
I already have and use pre-existing & still valid long-lived access token ;)
